File: 0486-predict-the-winner/0486-predict-the-winner.cpp
class Solution {
public:
    bool PredictTheWinner(vector<int>& nums) {
     int n = nums.size();
        vector<vector<int>> dp(n,vector<int>(n));
        for (int i = 0; i < n; ++i) {
            dp[i][i] = nums[i];
        }
        
        for (int diff = 1; diff < n; ++diff) {
            for (int left = 0; left < n - diff; ++left) {
                int right = left + diff;
                dp[left][right] = max(nums[left] - dp[left + 1][right],
                                          nums[right] - dp[left][right - 1]);
            }
        }
        
        return dp[0][n - 1] >= 0;   
    }
};

File: 0486-predict-the-winner/0486-predict-the-winner.java
class Solution {
    private int helper(int[] nums, int left, int right) {
        if (left == right) return nums[left];
        return Math.max(nums[left] - helper(nums, left + 1, right), 
                        nums[right] - helper(nums, left, right - 1));
    }
    public boolean predictTheWinner(int[] nums) {
        return helper(nums, 0, nums.length - 1) >= 0;
    }
}

/*
Complexity:
Time: O(2^n)
Space: O(n) (recursion stack depth)
*/